### PR TITLE
Make sure derived table column names are handled correctly

### PR DESCRIPTION
**Pull request by systay**
## Description
For some situations, during planning we create derived tables. When this happens, we sometimes get the column names wrong, which leads to queries that are not correct.

Example query, with bad plan:

```sql
select id, id from user 
union 
select id, bar from user_extra 
order by 1
```

bad plan produced:

```json
{
  "QueryType": "SELECT",
  "Original": "select id, id from user union select id, bar from user_extra order by 1",
  "Instructions": {
    "OperatorType": "Sort",
    "Variant": "Memory",
    "OrderBy": "(2|3) ASC",
    "ResultColumns": 2,
    "Inputs": [
      {
        "OperatorType": "Distinct",
        "Collations": [
          "(0:4)",
          "(1:4)",
          "(2:3)",
          "3"
        ],
        "Inputs": [
          {
            "OperatorType": "Route",
            "Variant": "Scatter",
            "Keyspace": {
              "Name": "user",
              "Sharded": true
            },
            "FieldQuery": "select dt.id, dt.id, id, weight_string(id), weight_string(dt.id) from (select id, id from `user` where 1 != 1 union select id, bar from user_extra where 1 != 1) as dt where 1 != 1",
            "Query": "select dt.id, dt.id, id, weight_string(id), weight_string(dt.id) from (select id, id from `user` union select id, bar from user_extra) as dt",
            "Table": "`user`, user_extra"
          }
        ]
      }
    ]
  },
  "TablesUsed": [
    "user.user",
    "user.user_extra"
  ]
}
```

As you can see from the plan, one of the queries being sent down to mysql looks like:

```sql
select dt.id, dt.id, id, weight_string(id), weight_string(dt.id)
from (
	select id, id from `user` 
	union 
	select id, bar from user_extra) as dt
```

This query would fail in MySQL with an ambiguous column error for `dt.id`.

Since this change is quite involved, and these queries have never worked well, I suggest we don't backport this change set.

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Is this aiming for the 19.0.2 release?

File: go/vt/vtgate/planbuilder/testdata/memory_sort_cases.json [hunk 1]
```diff
@@ -655,5 +655,55 @@
...
+      "QueryType": "SELECT",
+      "Original": "select * from (select u.foo, ue.bar from user u, user_extra ue) tbl order by tbl.bar, tbl.foo",
+      "Instructions": {
+        "OperatorType": "Sort",
+        "Variant": "Memory",
+        "OrderBy": "(1|2) ASC, (0|3) ASC",
+        "ResultColumns": 2,
+        "Inputs": [
+          {
+            "OperatorType": "Join",
```
Comment (systay) [inline go/vt/vtgate/planbuilder/testdata/memory_sort_cases.json:672]: nitpicking here, but a better plan would push the `tbl.bar` column to the LHS, and then we only need to do memory sort on the inputs from the RHS. We would not need to sort the full result set with that plan

**Comment by systay** [comment]
Still not done with the changes around this logic, but this is a big change set already, so I thought we could review and merge it as it currently stands. 

More to come soon

File: go/vt/vtgate/planbuilder/operators/aggregator.go [hunk 2]
```diff
@@ -190,6 +190,66 @@ func (a *Aggregator) AddColumn(ctx *plancontext.PlanningContext, reuse bool, gro
...
+		for _, aggr := range a.Aggregations {
+			if aggr.ColOffset != offset {
+				continue
+			}
+			if aggr.WSOffset >= 0 {
+				// ah, we already have a weigh_string for this column. let's return it as is
+				return aggr.WSOffset
+			}
+
+			panic(vterrors.VT13001("did not expect to get here"))
```
Comment (GuptaManan100) [inline go/vt/vtgate/planbuilder/operators/aggregator.go:225]: Maybe a better error message in case we do run into this in the future?

**Review by GuptaManan100 — APPROVED** [review]
Rest looks good to me!

File: go/vt/vtgate/planbuilder/operators/projection_pushing.go [hunk 3]
```diff
@@ -255,23 +260,16 @@ func splitProjectionAcrossJoin(
...
-	case Offset:
+	case Offset, nil:
 		// for offsets, we'll just treat the expression as unexplored, and later stages will handle the new offset
-		join.JoinColumns.add(splitUnexploredExpression(ctx, join, lhs, rhs, pe, colAlias))
+		join.JoinColumns.add(splitUnexploredExpression(ctx, join, lhs, rhs, pe, colAlias, dt))
 	case SubQueryExpression:
 		join.JoinColumns.add(splitSubqueryExpression(ctx, join, lhs, rhs, pe, colAlias))
 	default:
-		panic(dbg.S(pe.Info))
+		panic(vterrors.VT13001(dbg.S(pe.Info)))
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/projection_pushing.go:272]: nit: we should add some text to understand the internal error better.

File: go/vt/vtgate/planbuilder/operators/projection_pushing.go [hunk 4]
```diff
@@ -292,9 +290,24 @@ func splitUnexploredExpression(
...
 	alias string,
+	dt *DerivedTable,
 ) applyJoinColumn {
+	original := sqlparser.CloneRefOfAliasedExpr(pe.Original)
+	expr := pe.ColExpr
+
+	if dt != nil {
+		if !pe.isSameInAndOut(ctx) {
+			panic("not sure what to do here")
+		}
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/projection_pushing.go:301]: nit: can change the error message here.

File: go/vt/vtgate/planbuilder/operators/route.go [hunk 5]
```diff
@@ -672,23 +672,81 @@ func addMultipleColumnsToInput(
...
 		return addMultipleColumnsToInput(ctx, proj, reuse, addToGroupBy, exprs)
 	default:
 		return "", op, false, nil
 	}
 }
 
+func (r *Route) AddWSColumn(ctx *plancontext.PlanningContext, offset int, _ bool) int {
+	columns := r.GetColumns(ctx)
+	if offset > len(columns) {
+		panic(vterrors.VT13001("column %d not found", offset))
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/route.go:685]: this does not accept `format`, should be a string

File: go/vt/vtgate/planbuilder/operators/route.go [hunk 6]
```diff
@@ -672,23 +672,81 @@ func addMultipleColumnsToInput(
...
+	return false, -1
+}
+
+func addDerivedProj(
+	ctx *plancontext.PlanningContext,
+	op Operator,
+) (projection *Projection) {
+	// create new table ID
+	tableID := semantics.SingleTableSet(len(ctx.SemTable.Tables))
+	ctx.SemTable.Tables = append(ctx.SemTable.Tables, nil)
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/route.go:723]: nit: more comments can be added here as adding `nil` to semtable tables is not clear. I understand the intention here but comments will help.

Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operators/route.go:723]: Ideally we can add the `DerivedTable` here which is created below.

File: go/vt/vtgate/planbuilder/testdata/tpch_cases.json [hunk 7]
```diff
@@ -768,24 +1183,15 @@
...
-    "plan": "VT13002: unexpected AST struct for query: o_year"
-  },
-  {
-    "comment": "TPC-H query 9",
-    "query": "select nation, o_year, sum(amount) as sum_profit from ( select n_name as nation, extract(year from o_orderdate) as o_year, l_extendedprice * (1 - l_discount) - ps_supplycost * l_quantity as amount from part, supplier, lineitem, partsupp, orders, nation where s_suppkey = l_suppkey and ps_suppkey = l_suppkey and ps_partkey = l_partkey and p_partkey = l_partkey and o_orderkey = l_orderkey and s_nationkey = n_nationkey and p_name like '%green%' ) as profit group by nation, o_year order by nation, o_year desc",
-    "plan": "VT13002: unexpected AST struct for query: nation"
-  },
   {
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/tpch_cases.json:1195]: was this test removal intentional?

Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/testdata/tpch_cases.json:1195]: I see this is the problem with the GitHub diff.

**Review by harshit-gangal — APPROVED** [review]
Looks good overall. some nits

File: go/vt/vtgate/planbuilder/operators/route.go (hunk 6 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/route.go:723]: the Derived table that we create here is of a very different type than what we store in the semantic state, and the info that we normally store there does not really make sense to store here. we just need an ID

**Comment by deepthi** [comment]
@systay is there a user-reported issue for this fix that is still open and should be closed?

**Comment by systay** [comment]
> @systay is there a user-reported issue for this fix that is still open and should be closed?

no, this was found during our own testing

**Pull request merged**